### PR TITLE
Fix bug 1171957 - Rewrite scraper to use section importers

### PR DESCRIPTION
**Pull request by jwhitlock**
Previously, a single PEG grammar and data processor (``mdn/scrape.py``) was used to extract data from an MDN page.  This PR changes this significantly.

It fixes these bugs:
* [bug 1171957](https://bugzilla.mozilla.org/show_bug.cgi?id=1171957) - Refactor scraper into sub-parsers
* [bug 1174808](https://bugzilla.mozilla.org/show_bug.cgi?id=1174808) - Text between "Browser compatiblilty" header and the compat table breaks import tool
* [bug 1175177](https://bugzilla.mozilla.org/show_bug.cgi?id=1175177) - Work correctly on pages with swapped compat and spec sections
* [bug 1183593](https://bugzilla.mozilla.org/show_bug.cgi?id=1183593) - Parse links in specification comments

The changes are done as a series of commits with passing tests, but it may be better to review it on a file-by-file basis:

* ``mdn/visitor.py`` - Base classes setup the interface for converting a parsed grammar (``Visitor``) and processing a converted document (``Extractor``)
* ``mdn/data.py`` - ``Data`` provides database lookup and caching services to the parsers, with the (remote) possibility of using the API instead of direct database access through Django models.
* ``mdn/html.py`` - Provides a grammar for parsing well-formed HTML body sections, which ``HTMLVisitor`` converts into elements representing the HTML parts.
* ``mdn/kumascript.py`` - Augments the HTML grammar with KumaScript macros, which ``KumaVisitor`` converts into elements representing the macros, their data, and their API representations.
* ``mdn/specifications.py`` - ``SpecSectionExtractor`` takes a Specifications section that has been converted into elements, and finds the elements representing specifications data.  It passes these to ``SpecNameVisitor``, ``Spec2Visitor``, and ``SpecDescVisitor``, which re-parse and extract the data.
* ``mdn/compatibility.py`` - ``CompatSectionExtractor`` takes a Browser Compatibility section that has been converted into elements, and finds the elements representing compatibility data.  These are re-parsed with further customized grammars, and passed to ``CompatFeatureVisitor``, ``CompatSupportVisitor``, and ``CompatFootnoteVisitor`` to extract the data.
* ``mdn/scrape.py`` - With most of the data extraction functionality moved to other files, ``scrape_page`` is left to convert a raw MDN page into elements, using ``PageExtractor`` to collect them into sections and pass the relevant ones to ``SpecSectionExtractor`` and ``CompatSectionExtractor``.

Other changes are included, such as moving issue definitions from ``mdn/model.py`` to ``mdn/issues.py``, and some documentation cleanup.

### CODE REVIEW

**Comment by jwhitlock** [comment]
Rebased on top of PR #37, which fixes an issue that would have prevented practical testing.  No more rebases expected.

**Comment by jwhitlock** [comment]
Here's the issue counts for the rewritten importer.  "Old Count" is the current importer running on [browsercompat.herokuapp.com](https://browsercompat.herokuapp.com/importer/issues), and "New Count" is the local import in my dev box.  Highlights:

* ``section_skipped``, ``section_missed``, and ``halt_import`` drop to 0.  The rewritten importer can continue parsing, but generates issues such as ``tag_dropped`` (unexpected HTML elements), ``skipped_content`` (such as paragraphs before a compat table), ``no_data`` (usually, whole document is wrapped in a ``<div>``) and ``skipped_h3`` (``<h3>`` after compat table) instead.
* Because more content is parsed, there are a lot more ``unknown_kumascript``, ``inline_text``, ``unexpected_attribute`` errors
* New ``unexpected_kumascript`` for macro use in unexpected sections

I'll start opening issues for the items that will require importer fixes, but won't start submitting code until this PR is merged.

        Issue Slug         | Old Count | New Count
---------------------------|-----------|----------
**Total**                  |      1977 |      5607
section_skipped            |       873 |         0
unknown_version            |       324 |       244
unknown_kumascript         |       246 |      1926
inline_text                |       190 |       771
halt_import                |       113 |         0
unexpected_attribute       |        72 |       229
footnote_no_id             |        32 |       122
tag_dropped                |        20 |       767
skipped_h3                 |        19 |        98
footnote_missing           |        19 |        74
section_missed             |        10 |         0
footnote_multiple          |        10 |        25
failed_download            |         8 |         8
unknown_browser            |         7 |        24
exception                  |         7 |         1
footnote_unused            |         6 |        65
span_dropped               |         5 |         0
compatgeckodesktop_unknown |         5 |         7
spec_h2_name               |         2 |         8
spec_h2_id                 |         2 |         9
spec2_converted            |         2 |         6
second_footnote            |         2 |         0
missing_attribute          |         2 |        21
specname_not_kumascript    |         1 |         9
skipped_content            |         0 |       478
feature_header             |         0 |         4
spec2_omitted              |         0 |         3
specname_omitted           |         0 |         1
specname_converted         |         0 |         1
footnote_feature           |         0 |        16
cell_out_of_bounds         |         0 |         1
unknown_spec               |         0 |         6
spec_mismatch              |         0 |         6
unexpected_kumascript      |         0 |       485
no_data                    |         0 |        63
kumascript_wrong_args      |         0 |       129

File: mdn/visitor.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        """
+        return state, True  # pragma: no cover
+
+    def leaving_element(self, state, element):
+        """Extract data (state or information) from leaving an element."""
+        return state  # pragma: no cover
+
+    def extracted_data(self):
+        """Finalize the extracted data."""
+        return {}  # pragma: no cover
```
Comment (groovecoder) [inline mdn/visitor.py]: change to throwing `NotImplemented` ?

**Comment by groovecoder** [comment]
@jwhitlock and I did a code walk-through today. Aside from this one comment, I agreed with the code and reasoning behind it. r+ after the `NotImplemented` fix.

**Pull request merged**